tests/basic.rs

```diff
@@ -22,11 +22,58 @@ fn meta() {
 }
 
 #[test]
-fn next_path() {
+fn path() {
     let mut s = String::new();
-    Settings::path(&mut [1, 0, 0].iter().copied(), &mut s, "/").unwrap();
+    assert_eq!(Settings::path([1, 0, 0], &mut s, "/"), Ok(1));
     assert_eq!(s, "/b");
     s.clear();
-    Settings::path(&mut [2, 0, 0].iter().copied(), &mut s, "/").unwrap();
+    assert_eq!(Settings::path([2, 0, 0], &mut s, "/"), Ok(2));
     assert_eq!(s, "/c/inner");
+    s.clear();
+    assert_eq!(Settings::path([2], &mut s, "/"), Err(Error::TooShort(1)));
+    assert_eq!(s, "/c");
+    s.clear();
+    assert_eq!(Option::<i8>::path([0; 0], &mut s, "/"), Ok(0));
+    assert_eq!(s, "");
+}
+
+#[test]
+fn indices() {
+    let mut s = [0usize; 2];
+    assert_eq!(Settings::indices(["b", "foo"], s.iter_mut()), Ok(1));
+    assert_eq!(s, [1, 0]);
+    assert_eq!(
+        Settings::indices(["c", "inner", "bar"], s.iter_mut()),
+        Ok(2)
+    );
+    assert_eq!(s, [2, 0]);
+    assert_eq!(
+        Settings::indices(["c"], s.iter_mut()),
+        Err(Error::TooShort(1))
+    );
+    assert_eq!(Option::<i8>::indices([0; 0], s.iter_mut()), Ok(0));
+}
+
+#[test]
+fn traverse_empty() {
+    #[derive(Miniconf, Default)]
+    struct S {}
+    let f = |_, _: &_| unreachable!();
+    assert_eq!(
+        S::traverse_by_key([0].into_iter(), f),
+        Err(Error::<()>::NotFound(1))
+    );
+    assert_eq!(
+        S::traverse_by_key([0; 0].into_iter(), f),
+        Err(Error::TooShort(0))
+    );
+    assert_eq!(Option::<i32>::traverse_by_key([0].into_iter(), f), Ok(0));
```

Comment:
Should `traverse_by_key()` and friends take an `IntoIter`?

Reply:
I thought long and hard and went back and forth on this. I don't think so:
For the recursive pattern (`*_by_key()`) we take Iterators, since IMO the code noise of doing `.into_iter()` at each level in the implementation outweighs the flexibility gained and outweighs the code noise of making the iterator at the callsite. Consequently the non-recursive methods that are much more frequently called by the user (`{get,set}_json()`, `path()`, `indices()`) do take `IntoIterator`.
In short it does seem like `IntoIterator` doesn't look so good for recursive consumption because of the pointless `into_iter()` at all lower recursion levels, while it is perfect for iterative consumption where you only do `into_iter()` once().